### PR TITLE
Enable privacy tests

### PR DESCRIPTION
**Pull request by paul-szczepanek-arm**
Privacy tests needed to be updated to use the new API and cli-app needed a new command to help with scanning for unknown addresses. Right now almost all tests fail but you can use this to test while privacy is being fixed.

### CODE REVIEW

File: test_suite/gap/test_gap_privacy.py [hunk 1]
```diff
@@ -73,640 +31,347 @@ def address_is_random_non_resolvable(address):
...
+    # central_ss should see pairing complete successfully too
+    central_ss.expect_pairing_success()
 
-    configure_peripheral(device, rand_data)
-    yield device
+    central.gap.disconnect(peripheral_ss.connection_handle, "USER_TERMINATION")
 
-    device.ble.shutdown()
-    board_allocator.release(device)
+    # FIXME: needs event from the peripheral and central that they have been disconnected
```
Comment (pan-) [inline test_suite/gap/test_gap_privacy.py:64]: Events are present in a queue in the `events` field. However no JSON deserialization has been applied. We should improve this.

File: test_suite/gap/test_gap_privacy.py [hunk 2]
```diff
@@ -73,640 +31,347 @@ def address_is_random_non_resolvable(address):
...
 
 
 @pytest.mark.ble42
-@pytest.mark.skip(reason="Missing BLE API")
-@pytest.mark.parametrize("advertising_type", [
-     "ADV_CONNECTABLE_UNDIRECTED",
-     # "ADV_CONNECTABLE_DIRECTED",   # Directed advertising not supported yet
-     "ADV_SCANNABLE_UNDIRECTED",
-     "ADV_NON_CONNECTABLE_UNDIRECTED"
+@pytest.mark.parametrize("advertising_type, unresolvable", [
```
Comment (pan-) [inline test_suite/gap/test_gap_privacy.py:74]: I think we should remove the ble42 marker of these tests.

File: test_suite/common/gap_utils.py [hunk 3]
```diff
@@ -200,37 +201,61 @@ def __init__(self):
...
+    peripheral_connection = peripheral_connection_cmd.result
+    return central_connection, peripheral_connection
+
 
-    server_connection = server_connection_cmd.result
+def gap_connect(central, peripheral):
+    advertising_data = start_advertising_of_type(peripheral, "CONNECTABLE_UNDIRECTED")
+
+    scan = start_scanning_for_data(central, advertising_data)
+    sleep(0.1)
```
[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply test_suite/common/gap_utils.py:248]: removed

File: ble-cliapp/source/Commands/GapCommands.cpp [hunk 4]
```diff
@@ -1080,6 +1080,125 @@ DECLARE_CMD(ScanForAddress) {
     };
 };
 
+DECLARE_CMD(ScanForData) {
```
[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply ble-cliapp/source/Commands/GapCommands.cpp:1083]: added

File: test_suite/gap/test_gap_privacy.py [hunk 5]
```diff
@@ -73,640 +31,347 @@ def address_is_random_non_resolvable(address):
...
-    return (
-        scan_records[0]["peerAddrType"],
-        scan_records[0]["peerAddr"]
-    )
+def perform_bonding_with_privacy_and_disconnect(peripheral, central):
+    central_ss, peripheral_ss = init_security_sessions(central, peripheral)
 
+    peripheral.gap.enablePrivacy(True)
+    central.gap.enablePrivacy(True)
+    sleep(0.1)
```
[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply test_suite/gap/test_gap_privacy.py:39]: added

File: test_suite/gap/test_gap_privacy.py (hunk 2 above)
Comment (paul-szczepanek-arm) [reply test_suite/gap/test_gap_privacy.py:73]: changed to ble41

File: test_suite/gap/test_gap_privacy.py (hunk 1 above)
Comment (paul-szczepanek-arm) [reply test_suite/gap/test_gap_privacy.py:63]: do you want me to do this as part of this jira task?

Comment (pan-) [reply test_suite/gap/test_gap_privacy.py:63]: Jira task is enough. I got a one line _fix_ for ble-cliapp to not output event in the middle of a command response. The task would be about adding JSON deserialization of events.

File: test_suite/common/gap_utils.py [hunk 6]
```diff
@@ -200,37 +201,60 @@ def __init__(self):
...
+    peripheral.advParams.setPrimaryInterval(100, 100)
+    peripheral.gap.setAdvertisingParameters(LEGACY_ADVERTISING_HANDLE)
+
+    # payload
     peripheral.advDataBuilder.clear()
-    # add flags data into advertising payload
-    advertising_flags = ["LE_GENERAL_DISCOVERABLE", "BREDR_NOT_SUPPORTED"]
-    for f in advertising_flags:
-        peripheral.advDataBuilder.setFlags(f)
+    peripheral.advDataBuilder.setFlags("LE_GENERAL_DISCOVERABLE")
```
Comment (pan-) [inline test_suite/common/gap_utils.py:212]: This flag should not be set if the advertising type is not connectable.

File: test_suite/common/gap_utils.py [hunk 7]
```diff
@@ -200,37 +201,60 @@ def __init__(self):
         self.size = self.dataSize + 2
 
 
-def gap_connect(central, peripheral):
+def start_advertising_of_type(peripheral, advertising_type):
+    # params
+    peripheral.advParams.setType(advertising_type)
+    peripheral.advParams.setPrimaryInterval(100, 100)
```
Comment (pan-) [inline test_suite/common/gap_utils.py:207]: Please provide other values for non connectable advertising, ST controllers have some difficulties with thos. 
I suggest `200, 400` when the advertising type is not connectable.

**Review by pan- — APPROVED** [review]

**Pull request merged**